File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/onboarding/StartupActivity.java
package de.tum.in.tumcampusapp.component.ui.onboarding;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.ContentLoadingProgressBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;

import com.crashlytics.android.Crashlytics;

import java.util.concurrent.atomic.AtomicBoolean;

import androidx.work.State;
import androidx.work.WorkStatus;
import de.tum.in.tumcampusapp.BuildConfig;
import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.api.app.AuthenticationManager;
import de.tum.in.tumcampusapp.component.ui.overview.MainActivity;
import de.tum.in.tumcampusapp.service.StartSyncReceiver;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.Utils;
import io.fabric.sdk.android.Fabric;

import static android.Manifest.permission.ACCESS_COARSE_LOCATION;
import static android.Manifest.permission.ACCESS_FINE_LOCATION;
import static android.content.pm.PackageManager.PERMISSION_GRANTED;

/**
 * Entrance point of the App.
 */
public class StartupActivity extends AppCompatActivity {

    private static final int REQUEST_LOCATION = 0;
    private static final String[] PERMISSIONS_LOCATION = {ACCESS_COARSE_LOCATION,
                                                          ACCESS_FINE_LOCATION};

    final AtomicBoolean initializationFinished = new AtomicBoolean(false);
    private int tapCounter; // for easter egg

    @Nullable
    LiveData<WorkStatus> downloadStatus;

    // TODO: register background download finished WorkStatus
    Observer<WorkStatus> downloadObserver = workStatus -> {
        if (workStatus == null || workStatus.getState() != State.SUCCEEDED) {
            return;
        }
        openMainActivityIfInitializationFinished();
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_startup);

        // Only use Crashlytics if we are not compiling debug
        if (!BuildConfig.DEBUG) {
            Fabric.with(this, new Crashlytics());
        }

        initEasterEgg();

        new Thread(this::init).start();
    }

    private void initEasterEgg() {
        if (Utils.getSettingBool(this, Const.RAINBOW_MODE, false)) {
            ImageView tumLogo = findViewById(R.id.startupTumLogo);
            tumLogo.setImageResource(R.drawable.tum_logo_rainbow);
        }

        tapCounter = 0;
        View background = findViewById(R.id.container);
        background.setOnClickListener(view -> {
            tapCounter++;
            if (tapCounter % 3 == 0) {
                tapCounter = 0;

                // use the other logo and invert the setting
                boolean rainbowEnabled = Utils.getSettingBool(this, Const.RAINBOW_MODE, false);
                rainbowEnabled = !rainbowEnabled;
                ImageView tumLogo = findViewById(R.id.startupTumLogo);

                if (rainbowEnabled) {
                    tumLogo.setImageResource(R.drawable.tum_logo_rainbow);
                } else {
                    tumLogo.setImageResource(R.drawable.tum_logo_blue);
                }

                Utils.setSetting(this, Const.RAINBOW_MODE, rainbowEnabled);
            }
        });
        background.setSoundEffectsEnabled(false);
    }

    private void init() {
        // Migrate all settingsPrefix - we somehow ended up having two different shared prefs: join them back together
        Utils.migrateSharedPreferences(this.getApplicationContext());

        // Check that we have a private key setup in order to authenticate this device
        AuthenticationManager am = new AuthenticationManager(this);
        am.generatePrivateKey(null);

        // On first setup show remark that loading could last longer than normally
        runOnUiThread(() -> {
            ContentLoadingProgressBar progressBar = findViewById(R.id.startupLoadingProgressBar);
            progressBar.show();
        });

        // TODO: manually start DownloadWorker to listen to success
        // Start background service and ensure cards are set
        Intent i = new Intent(this, StartSyncReceiver.class);
        i.putExtra(Const.APP_LAUNCHES, true);
        sendBroadcast(i);

        // Request Permissions for Android 6.0
        requestLocationPermission();
    }

    /**
     * Request the Location Permission
     */
    private void requestLocationPermission() {
        //Check, if we already have permission
        if (ContextCompat.checkSelfPermission(this, ACCESS_COARSE_LOCATION) == PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(this, ACCESS_FINE_LOCATION) == PERMISSION_GRANTED) {
            // We already got the permissions, to proceed normally
            openMainActivityIfInitializationFinished();
        } else if (ActivityCompat.shouldShowRequestPermissionRationale(this, ACCESS_COARSE_LOCATION) ||
                ActivityCompat.shouldShowRequestPermissionRationale(this, ACCESS_FINE_LOCATION)) {
            // Provide an additional rationale to the user if the permission was not granted
            // and the user would benefit from additional context for the use of the permission.
            // For example, if the request has been denied previously.

            // Display an AlertDialog with an explanation and a button to trigger the request.
            runOnUiThread(() -> {
                AlertDialog dialog = new AlertDialog.Builder(this)
                        .setMessage(getString(R.string.permission_location_explanation))
                        .setPositiveButton(R.string.ok, (dialogInterface, id) -> {
                            ActivityCompat.requestPermissions(
                                    this, PERMISSIONS_LOCATION, REQUEST_LOCATION);
                        })
                        .create();

                if (dialog.getWindow() != null) {
                    dialog.getWindow().setBackgroundDrawableResource(
                            R.drawable.rounded_corners_background);
                }

                dialog.show();
            });
        } else {
            ActivityCompat.requestPermissions(this, PERMISSIONS_LOCATION, REQUEST_LOCATION);
        }
    }

    /**
     * Callback when the user allowed or denied Permissions
     * We do not care, if we got the permission or not, since the LocationManager needs to handle
     * missing permissions anyway
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        openMainActivityIfInitializationFinished();
    }

    private void openMainActivityIfInitializationFinished() {
        if (initializationFinished.compareAndSet(false, true) || isFinishing()) {
            // If the initialization process is not yet finished or if the Activity is
            // already being finished, there's no need to open MainActivity.
            return;
        }
        openMainActivity();
    }

    /**
     * Animates the TUM logo into place (left upper corner) and animates background up.
     * Afterwards {@link MainActivity} gets started
     */
    private void openMainActivity() {
        Intent intent = new Intent(StartupActivity.this, MainActivity.class);
        startActivity(intent);
        finish();
        overridePendingTransition(R.anim.fadein, R.anim.fadeout);
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/onboarding/StartupActivity.kt
package de.tum.`in`.tumcampusapp.component.ui.onboarding

import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat.checkSelfPermission
import androidx.core.app.ActivityCompat.requestPermissions
import androidx.core.app.ActivityCompat.shouldShowRequestPermissionRationale
import androidx.lifecycle.LiveDataReactiveStreams
import com.crashlytics.android.Crashlytics
import de.tum.`in`.tumcampusapp.BuildConfig.DEBUG
import de.tum.`in`.tumcampusapp.BuildConfig.VERSION_CODE
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.api.app.AuthenticationManager
import de.tum.`in`.tumcampusapp.api.tumonline.CacheControl
import de.tum.`in`.tumcampusapp.component.other.generic.activity.BaseActivity
import de.tum.`in`.tumcampusapp.component.other.generic.activity.BaseNavigationActivity
import de.tum.`in`.tumcampusapp.service.DownloadWorker
import de.tum.`in`.tumcampusapp.service.StartSyncReceiver
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.Utils
import de.tum.`in`.tumcampusapp.utils.observe
import io.fabric.sdk.android.Fabric
import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_startup.container
import kotlinx.android.synthetic.main.activity_startup.startupLoadingProgressBar
import kotlinx.android.synthetic.main.activity_startup.startupTumLogo
import org.jetbrains.anko.doAsync
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject

class StartupActivity : BaseActivity(R.layout.activity_startup) {

    private val initializationFinished = AtomicBoolean(false)
    private var tapCounter = 0 // for easter egg

    @Inject
    lateinit var workerActions: DownloadWorker.WorkerActions

    @Inject
    lateinit var authManager: AuthenticationManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.downloadComponent().inject(this)

        // Only use Crashlytics if we are not compiling debug
        val isDebuggable = applicationInfo.isDebuggable
        if (!DEBUG && !isDebuggable) {
            Fabric.with(this, Crashlytics())
            Crashlytics.setString("TUMID", Utils.getSetting(this, Const.LRZ_ID, ""))
            Crashlytics.setString("DeviceID", AuthenticationManager.getDeviceID(this))
        }

        val savedAppVersion = Utils.getSettingInt(this, Const.SAVED_APP_VERSION, VERSION_CODE)
        if (savedAppVersion < VERSION_CODE) {
            Utils.setSetting(this, Const.SHOW_UPDATE_NOTE, true)
            Utils.setSetting(this, Const.UPDATE_MESSAGE, "")
        }
        // Always set current app version, otherwise it will never be initialized and the update
        // note is never displayed
        Utils.setSetting(this, Const.SAVED_APP_VERSION, VERSION_CODE)

        initEasterEgg()
        doAsync {
            initApp()
        }
    }

    private fun initEasterEgg() {
        if (Utils.getSettingBool(this, Const.RAINBOW_MODE, false)) {
            startupTumLogo.setImageResource(R.drawable.tum_logo_rainbow)
        }

        container.setOnClickListener {
            if (tapCounter++ % 3 == 0) {
                // Switch to the other logo and invert the setting
                val shouldEnableRainbow = Utils.getSettingBool(this, Const.RAINBOW_MODE, false).not()

                if (shouldEnableRainbow) {
                    startupTumLogo.setImageResource(R.drawable.tum_logo_rainbow)
                } else {
                    startupTumLogo.setImageResource(R.drawable.tum_logo_blue)
                }

                Utils.setSetting(this, Const.RAINBOW_MODE, shouldEnableRainbow)
            }
        }
    }

    private fun initApp() {
        // Migrate all settings - we somehow ended up having two different shared prefs: join them
        // back together
        Utils.migrateSharedPreferences(this)

        // Check that we have a private key setup in order to authenticate this device
        authManager.generatePrivateKey(null)

        // On first setup show remark that loading could last longer than normally
        runOnUiThread {
            startupLoadingProgressBar.show()
        }

        // Start download workers and listen for finalization
        val downloadActions = Flowable
            .fromCallable(this::performAllWorkerActions)
            .onErrorReturnItem(Unit)
            .subscribeOn(Schedulers.io())

        runOnUiThread {
            LiveDataReactiveStreams
                .fromPublisher(downloadActions)
                .observe(this) { openMainActivityIfInitializationFinished() }
        }

        // Start background service and ensure cards are set
        sendBroadcast(Intent(this, StartSyncReceiver::class.java))

        // Request permissions for Android 6 and up
        requestLocationPermission()
    }

    private fun performAllWorkerActions() {
        for (action in workerActions.actions) {
            action.execute(CacheControl.USE_CACHE)
        }
    }

    private fun requestLocationPermission() {
        when {
            hasLocationPermissions -> openMainActivityIfInitializationFinished()
            shouldShowRationale -> runOnUiThread { showLocationPermissionRationaleDialog() }
            else -> requestPermissions(this, PERMISSIONS_LOCATION, REQUEST_LOCATION)
        }
    }

    private val hasLocationPermissions: Boolean
        get() = PERMISSIONS_LOCATION.all { checkSelfPermission(this, it) == PERMISSION_GRANTED }

    private val shouldShowRationale: Boolean
        get() = PERMISSIONS_LOCATION.all { shouldShowRequestPermissionRationale(this, it) }

    /**
     * Displays a dialog to the user explaining why we need the location permissions
     */
    private fun showLocationPermissionRationaleDialog() {
        AlertDialog.Builder(this, R.style.Theme_MaterialComponents_Light_Dialog)
            .setMessage(R.string.permission_location_explanation)
            .setPositiveButton(R.string.ok) { _, _ ->
                requestPermissions(this, PERMISSIONS_LOCATION, REQUEST_LOCATION)
            }
            .create()
            .apply {
                window?.setBackgroundDrawableResource(R.drawable.rounded_corners_background)
            }
            .show()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        openMainActivityIfInitializationFinished()
    }

    private fun openMainActivityIfInitializationFinished() {
        if (initializationFinished.compareAndSet(false, true) || isFinishing) {
            // If the initialization process is not yet finished or if the Activity is
            // already being finished, there's no need to open MainActivity.
            return
        }
        openMainActivity()
    }

    private fun openMainActivity() {
        val intent = Intent(this, BaseNavigationActivity::class.java)
        startActivity(intent)
        finish()
        overridePendingTransition(R.anim.fadein, R.anim.fadeout)
    }

    private val ApplicationInfo.isDebuggable: Boolean
        get() = 0 != (flags and ApplicationInfo.FLAG_DEBUGGABLE)

    private companion object {
        private const val REQUEST_LOCATION = 0
        private val PERMISSIONS_LOCATION = arrayOf(ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION)
    }
}
